### PR TITLE
[Cluster launcher] [Azure] Make cluster termination and networking more configurable and robust

### PR DESCRIPTION
**Pull request by mjd3**
## Why are these changes needed?
This PR addresses a few issues when launching clusters with Azure:
1. Any changes made to subnets of the deployed virtual network(s) are bashed upon redeployment.
    -  Any service endpoints, route tables, or delegations are removed when redeploying (which happens on any of the `ray` CLI calls) due to [this open Azure issue](https://github.com/Azure/azure-quickstart-templates/issues/2786). This PR provides a workaround for the issue by copying the existing subnet configuration into the deployment template if a subnet already exists with the cluster unique id within the same resource group.
2. VM termination is extremely lengthy and does not clean up all dependencies.
    - When VMs are provisioned, dependencies such as disks, NICs, and public IP addresses are also provisioned. However, because the termination process does not wait for the VM to be deleted and the dependent resources cannot be deleted at the same time as the VM, these dependencies are often left in the resource group after termination. This can cause issues with quotas (i.e., reaching a limit of public IP addresses or disks) and wastes resources. This PR moves node termination into a pool of threads so that node deletion can be parallelized (since waiting for each node to be deleted takes a long time) and all dependencies can be correctly deleted once their VMs no longer exist.
3. VMs can have status code `ProvisioningState/failed/RetryableError`, causing an unpacking error.
    - [This line](https://github.com/ray-project/ray/blob/187a5c03ced8ce8988ef130a3ee7e4dd1b38b918/python/ray/autoscaler/_private/_azure/node_provider.py#L100) throws an exception when the provisioning state is the string above, resulting in incorrect provisioning/termination of the node. This PR addresses that issue by slicing the list of status strings and only using the first two.
4. The default quota for public IP addresses in Azure is only 100, which can result in quota limits being hit for larger clusters.
    - This PR adds an option (`use_external_head_ip`) for only provisioning a public IP address for the head node (instead of all nodes or no nodes). This allows a user to still communicate with the head node via a public IP address without running into quota limits on public IP addresses. This option works in tandem with `use_internal_ips` - if both are set to `True`, then a public IP address will only be provisioned for the head node. If `use_external_head_ip` is omitted, the behavior is unchanged from the current behavior (i.e., public IPs will be provisioned for all nodes if `use_internal_ips` is `False`, otherwise no public IPs will be provisioned).

I've tested all of these fixes using `ray up`/`ray dashboard`/`ray down` on Azure clusters of 4-32 nodes to make sure the start up/teardown works correctly and the correct amount of resources are provisioned.

## Related issue number
Node termination times are discussed in https://github.com/ray-project/ray/issues/25971

## Checks
- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [x] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206743421817428

### CODE REVIEW

File: python/ray/autoscaler/_private/_azure/node_provider.py
Comment (gramhagen) [inline python/ray/autoscaler/_private/_azure/node_provider.py:104]:
```suggestion
        self.terminating_nodes = {k: v for k, v in self.terminating_nodes.items() if not v.done()}
```

File: python/ray/autoscaler/_private/_azure/node_provider.py
Comment (gramhagen) [inline python/ray/autoscaler/_private/_azure/node_provider.py:429]: is it possible that this could hang if the deletion fails?

File: python/ray/autoscaler/_private/_azure/node_provider.py
Comment (gramhagen) [inline python/ray/autoscaler/_private/_azure/node_provider.py:452]: same question as above, maybe an upper limit on time to wait would be useful?

File: python/ray/autoscaler/_private/_azure/node_provider.py
Comment (gramhagen) [inline python/ray/autoscaler/_private/_azure/node_provider.py:459]:
```suggestion
        return self.cached_nodes.get(node_id, self._get_node(node_id=node_id))
```

**Review by gramhagen — APPROVED** [review]
Looks great! just some syntax suggestions and question on while loop limits. I'll leave it to you how to address, functionally looks good.

**Comment by mjd3** [comment]
Thanks for the suggestions @gramhagen; added some timeouts to those whiles just in case. @architkulkarni @hongchaodeng let me know if you have any other suggestions!

File: python/ray/autoscaler/_private/commands.py
Comment (architkulkarni) [inline python/ray/autoscaler/_private/commands.py:608]: Nit: I'm not sure why it was written like this... Is it possible it's to guard against something like the string `"False"` being truthy in Python, or any other nonempty value or typo in the config? What do you think?

**Review by architkulkarni — COMMENTED** [review]
Just reviewed the code in `commands.py` and `updater.py`, looks good. Deferring to @gramhagen's approval for the Azure-specific part.

Another nit: the check for external_head_ip is a bit complex and in a few places, would it make sense to factor it out and add a unit test for it?

File: python/ray/autoscaler/_private/commands.py
Comment (mjd3) [reply python/ray/autoscaler/_private/commands.py:608]: Yeah I was also confused by that; it's inconsistent within that file which is why I had changed it when adding the additional check for the head node external IP param. For example, [this line](https://github.com/ray-project/ray/blob/7c1ef5e5be4d3427706abf7387aabafc95ec1873/python/ray/autoscaler/_private/commands.py#L1361). So either way, one of these lines should likely be changed. Let me know what you think.

**Comment by mjd3** [comment]
> Another nit: the check for external_head_ip is a bit complex and in a few places, would it make sense to factor it out and add a unit test for it?

Happy to add a unit test for this; would you be able to point me to where that would belong? I didn't see any Azure tests [here](https://github.com/ray-project/ray/tree/master/python/ray/tests) and [this folder](https://github.com/ray-project/ray/tree/master/python/ray/tests/autoscaler) only contains one `utils.py` file.

File: python/ray/autoscaler/_private/_azure/node_provider.py
Comment (mjd3) [reply python/ray/autoscaler/_private/_azure/node_provider.py:478]: @gramhagen I actually had to revert this suggestion as it results in invalidating the cache every time (since `self._get_node(node_id=node_id)` is always called). The `or` allows for short-circuiting that call.

Comment (gramhagen) [reply python/ray/autoscaler/_private/_azure/node_provider.py:478]: oops, good catch!

**Comment by mjd3** [comment]
@architkulkarni following up on the above; let me know what next steps there are for this PR if any.

**Review by architkulkarni — APPROVED** [review]
Looks good, my comments were minor and don't need to block the PR.

**Pull request merged**
